ecosystem/sep-0012.md

```diff
@@ -49,10 +49,15 @@ Clients should submit the JWT previously obtained from the anchor via the [SEP-1
 Authorization: Bearer <JWT>
 ```
 
-Alternatively, if the client cannot add the authorization header. The JWT should be passed as a jwt query parameter:
-```
-?jwt=<token>
-```
+### Shared Accounts
+
+Client applications can use a single Stellar account to hold multiple customers' funds. To differentiate customers using the same Stellar account, the decoded SEP-10 JWT may contain `memo` and `memo_type` attributes. The anchor should use these attributes in addition to the Stellar account in their data model to represent unique customers. See the [SEP-10 Memos](sep-0010.md#memos) section for more information.
+
+Note that this scenario is not applicable to [SEP-31](sep-0031.md)'s use of this protocol. This is because the Stellar account used to authenticate in this case identifies the [Sending Anchor](sep-0031.md#abstract) itself, not a particular customer, and therefore the scope of the authenticated session should include all customers registered and transactions initiated by the Sending Anchor.
+
+To facilitate this, SEP-31 Sending Anchors should use the `memo` and `memo_type` request parameters described for the endpoints below to represent unique customers instead of using SEP-10 memos. 
+
+SEP-12 implementations should expect SEP-10 memos, if present in the decoded JWT, to match the memo request parameters and should return an error response if they do not.
```

Comment:
❗ I don't think we should say that rules apply to some SEPs and not others. SEP-12 needs to stand on its own with a single set of rules that can be applied uniformly without conditions on the SEP it is being used with, otherwise I think it is confusing to implement. I think we can tweak the language below to get the same result without presenting it as SEP-31 gets an exception to a rule.

```suggestion
SEP-12 implementations should expect memo claims in the SEP-10 JWT to match the memo request parameters and should return an error response if they do not. If the JWT does not contain memo claims then the memo request parameters may contain any value. This behavior allows shared account owners, such as [SEP-31 sending anchors](sep-0031.md#abstract), to submit user information about their users using the memos assigned to their users.
```

Thoughts?

---

Note that this section appears to answer my earlier question of what it means if no memo is provided: authentication with a parent account should provide access to all data of sub-accounts identified by memos. It may be worth calling this out explicitly in SEP-10, or at least specifying in SEP-10 that it is undefined behavior and clients cannot rely on the outcome in SEPs that use SEP-10.

Reply:
> authentication with a parent account should provide access to all data of sub-accounts identified by memos.

I have concerns about this approach, both from a security and application-logic perspective.

If a client application authenticates with an account that was previously used with memos, the session should be treated as disjoint from other sessions using memos with the same account.

This ensures the service never exposes transaction or KYC information that was not explicitly requested by the client using the composite primary key of (account, memo, memo_type).

This simplified application logic as well. Services can always use the three claims to retrieve user and transaction records instead of sometimes using memos and sometimes not.
```
# .get() either returns the value or None

def create_user(jwt):
    User.objects.create(account=jwt.get("account"), memo=jwt.get("memo"), memo_type=jwt.get("memo_type"))

def get_user(jwt):
    User.objects.get(account=jwt.get("account"), memo=jwt.get("memo"), memo_type=jwt.get("memo_type"))
```

Reply:
I agree with the text in the suggested change though, so I'll commit it but leave this thread unresolved.

Reply:
> I have concerns about this approach, both from a security and application-logic perspective.
> 
> If a client application authenticates with an account that was previously used with memos, the session should be treated as disjoint from other sessions using memos with the same account.

This is my preference and I share the same concern, but it sounds like the way it works with SEP-31 is that it isn't disjoint. I think we should endeavor to have our SEPs use SEP-10's scoping in the same way. How the scoping occurs is really out of scope of SEP-10 itself, but some consistency would have value.

Maybe we say in SEP-10 that it is not recommended, but then SEP-31/12 does it anyway?

Reply:
I'm not sure I understand how SEP-31's use of SEP-10 does not make sessions using the same account disjoint. In SEP-31's use of SEP-10, Sending Anchors will never request memos to be in their challenge transactions and downstream JWTs, so the Sending Anchor's session will always be disjoint from transactions and KYC data created in sessions authenticated by JWTs including memos. Let me know if you think I'm missing something.